### ISSUE TITLE
NERTagger.tag_sents() no longer tokenizes sentences.

### ISSUE THREAD

**Issue by tfry-git** [nltk/nltk#1007]
Apologies for double posting: I also posted this as a comment to commit c6b5fa6dcc360c6542328071c92ac666a70e195b but I'm not sure it gets read, there:

This commit seems to fix the problem twice:
1) Changing tag() to work ok, if the input is tokenized into multiple sentences, and
2) changing _cmd() to suppress sentence tokenization, entirely.

The latter causes a regression in tag_sents(). Now, this will return all sentences in a single element, too. I don't think this is intentional (and I for one clearly don't like it). I removed the addition of

```
'-tokenizerFactory', 'edu.stanford.nlp.process.WhitespaceTokenizer', '-tokenizerOptions','\"tokenizeNLs=false\"'
```

locally, and that restores the old behavior of tag_sents() for me.

Regards
Thomas

**Comment by longdt219** [comment]
Hi @tfry-git, 
Thank you very much for reporting this. This is however, a known bug #761. The problem is, currently there is no way to force Stanford NER to stop sentence tokenization because our input is assumed to be tokenized. Removing

-tokenizerFactory', 'edu.stanford.nlp.process.WhitespaceTokenizer', '-tokenizerOptions','\"tokenizeNLs=false\"

Will fail the following cases: 
(1) When sentence contains splitting characters. 
Eg. StandordNER.tag("Hello? Melbourne".split())

[[(u'Hello', u'PERSON'), (u'?', u'O')], [(u'Melbourne', u'LOCATION')]] 
So that length of input (=2) is different with the length of output (=3) 
(2) For input that contains more than 1 sentence 
Eg. nt.tag_sents(['I am going to fly to Paris tomorrow. Is Pierre going to come with me?'.split(),'I dont think so'.split()])
will return 3 sentences instead of 2 sentences.

I reported the bug to the Stanford dev mail list here and they said they are looking at it. However, it seems that they haven't done anything. 
Hi @stevenbird, probably we should do some hack to get over this i.e. to split the result according to the input. This will fix the bug for now. What do you think ?

**Comment by nltk-webdeploy** [comment]
@longdt219 yes, please implement a temporary workaround.

**Comment by tfry-git** [comment]
Hi!

Ok, I guess it was somewhat unclear to me what tag_sents() is supposed to do, then. Since I had developed my script with 3.0.0, originally, I was under the impression that segmentation/tokenization was always going to be done by the stanford code, but tag_sents() was simply the variant suited for tagging multiple sentences.

Actually, I found that behavior to be fairly useful. If it is not the intended default behavior, I would certainly suggest to make it available, as optional behavior. (Same for POSTagger, actually.)

Note that using the stanford tokenizer also provided robustness against certain types of easy to run-into problems, even for the single-sentence case. For instance, the stanford NER tagger expects punctuation to be provided as separate tokens (if input is pre-tokenized), which is typically not quite as easy to achieve as splitting raw input on whitespace. As one example of many, the code provided in issue #761 is incorrect in this case, in providing tokens such as "abend." containing punctuation. Having the stanford tagger deal with all such subtleties, appropriately and automatically, is definitely comfortable.

So, summary: This is not the same request as #761 . Rather I would like to (optionally) get back the behavior that I can safely pass non-tokenized (or mis-tokenized) text to the NER tagger.

Regards
Thomas

**Comment by longdt219** [comment]
Hi @tfry-git, 
I got what you mean. The input for NER tagger is presumed to be tokenized. However, you are free to use what ever tokenizer / sentence splitter here. You can achieve the same old behavior by using Stanford tokenizer first and then apply NER tagger.

**Comment by tfry-git** [comment]
Hi @longdt219,

true, I could chain the Stanford tokenizer and Stanford tagger. And if you don't like my suggestion, I'm not going to engage you in a lengthy back-and-forth. However, a few points to defend my suggestion:
1. While it is important to be able to make the Stanford tools work on existing tokenization / segmentation, a rather typical use case (in particular for the NER tagger) is to generate terms of interest from raw input. This use-case will be much easier and safer to implement, if the API supports passing raw input.
2. I suspect the reason why the tagger API does not support this in the first place is that other tagging tools do not come with the ability to take care of appropriate tokenization. The Stanford tools do, and this feature can be put to good use. And making this existing feature available in the API is rather easy.
3. In particular for the java based Stanford tools, there is a considerable performance penalty for every call to these tools. Having to make separate calls to Stanford tokenizer and Stanford tagger is painful, performance wise.

Regards
Thomas

**Comment by longdt219** [comment]
Hi @tfry-git, 
Thank you for your reply. Now I see why raw input is important. I propose that we keep the interface TaggerI (since it's shared for different taggers) but add methods to do raw tagging for stanford POS and NER. What do you think @stevenbird

**Comment by stevenbird** [comment]
Resolved by #1008.
